operator/pkg/lbipam/lbipam.go

```diff
@@ -0,0 +1,1797 @@
+
+	err := ipam.handleUpsertService(svc)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handleUpsertService()")
+		return fmt.Errorf("handleUpsertService: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) svcOnDelete(k resource.Key, svc *slim_core_v1.Service) error {
+	// Deep copy so we get a version we are allowed to update
+	svc = svc.DeepCopy()
+
+	ipam.logger.Debugf("Deleted service '%s/%s'", svc.GetNamespace(), svc.GetName())
+
+	err := ipam.handleDeletedService(svc)
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in handleDeletedService()")
+		return fmt.Errorf("handleDeletedService: %w", err)
+	}
+
+	err = ipam.satisfyAndUpdateCounts()
+	if err != nil {
+		ipam.logger.WithError(err).Error("Error in satisfyAndUpdateCounts()")
+		return fmt.Errorf("satisfyAndUpdateCounts: %w", err)
+	}
+
+	return nil
+}
+
+func (ipam *LBIPAM) satisfyAndUpdateCounts() error {
+	err := ipam.satisfyServices()
+	if err != nil {
+		return fmt.Errorf("satisfyServices: %w", err)
+	}
+
+	err = ipam.updateAllPoolCounts()
+	if err != nil {
+		return fmt.Errorf("updateAllPoolCounts: %w", err)
+	}
+
+	return nil
+}
+
+// handleUpsertService updates the service view in the service store, it removes any allocation and ingress that
+// do not belong on the service and will move the service to the satisfied or unsatisfied service view store depending
+// on if the service requests are satisfied or not.
+func (ipam *LBIPAM) handleUpsertService(svc *slim_core_v1.Service) error {
+	key := resource.NewKey(svc)
+	sv, found, _ := ipam.serviceStore.GetService(key)
+	if !found {
+		sv = &ServiceView{
+			Key: key,
+		}
+	}
+
+	// Ignore services which are not meant for us
+	if !ipam.isResponsibleForSVC(svc) {
+		if !found {
+			return nil
+		}
+
+		// Release allocations
+		for _, alloc := range sv.AllocatedIPs {
+			err := alloc.Origin.allocRange.Release(alloc.IP)
+			if err != nil {
+				return fmt.Errorf("alloc range release: %w", err)
+			}
+		}
+		ipam.serviceStore.Delete(sv.Key)
+
+		// Remove all ingress IPs
+		sv.Status.LoadBalancer.Ingress = nil
+		for i := len(sv.Status.Conditions) - 1; i >= 0; i-- {
+			if sv.Status.Conditions[i].Type == ciliumSvcRequestSatisfiedCondition {
+				sv.Status.Conditions = slices.Delete(sv.Status.Conditions, i, i+1)
+			}
+		}
+
+		err := ipam.patchSvcStatus(sv)
+		if err != nil {
+			return fmt.Errorf("patchSvcStatus: %w", err)
+		}
+
+		return nil
+	}
+
+	// We are responsible for this service.
+
+	// Update the service view
+	sv.Generation = svc.Generation
+	sv.Labels = svcLabels(svc)
+	sv.RequestedFamilies.IPv4, sv.RequestedFamilies.IPv6 = ipam.serviceIPFamilyRequest(svc)
+	sv.RequestedIPs = getSVCRequestedIPs(svc)
+	sv.Status = svc.Status.DeepCopy()
```

Comment:
Because we are modifying `sv` here

Reply:
We need to deep copy the status here since the object and its status are owned by the store. We do modify this status object so we can later patch it, this would cause the shared object in the store to change, which isn't allowed to avoid strange bugs with other components that also use the status.

Reply:
Modifying `sv` isn't an issue.